### PR TITLE
add Nonbacktracking type

### LINKED ISSUE #209: Nonbacktracking operator without explicit construction
Many things the non-backtracking operator is used for can be done with iterative solvers instead of direct solvers. I am thinking of detecting graph structures from the eigenvalues and eigenvectors or from solutions to `Mx=b` where `M` is the non-backtracking (NBT) operator. @CarloLucibello are you planning to do something with the NBT operator that needs the explicit matrix? Since some things can be done without it, we should add a function to return a `GraphMatrices.GraphMatrix` that represents the NBT operator. This would reduce storage costs from `nv(g)ne(g)` to `2ne(g)` and allow iterative solvers to answer these questions.

### PR DESCRIPTION
**Pull request by jpfairbanks**
Changes the community detection with nonbacktring operator
to use an implicit representation of the NBT operator.
This should allow for improved scalability since non_backtracking_matrix is dense.

The representation is based on GraphMatrices.jl

Tests are included comparing to old implementation on a small graph.

closes #209

### CODE REVIEW

**Comment by jpfairbanks** [comment]
I also moved the nonbacking code in src/spectral.jl to the bottom of the file because it less commonly used than Adjacency, Laplacian, or Incidence matrix.

**Comment by sbromberger** [comment]
So, two things:

* 1 line missing from coverage (https://codecov.io/github/JuliaGraphs/LightGraphs.jl/src/spectral.jl?ref=c5bffdcb6221b5bbc0e28b32fb41317b67e222bd#l-199).

* if you submit a new PR we should have the build issues on OSX fixed (I disabled them temporarily until Julia can get an OSX build working on Travis).

File: src/spectral.jl [hunk 1]
```diff
@@ -167,3 +129,134 @@ function incidence_matrix(g::SimpleGraph, T::DataType=Int)
...
+        for e in edges(g)
+            m += 1
+            edgeidmap[reverse(e)] = m
+        end
+    end
+    return Nonbacktracking(g, edgeidmap,m)
+end
+
+import Base.*
+function *{G, T<:Number}(nbt::Nonbacktracking{G}, x::Vector{T})
```
Comment (sbromberger) [inline src/spectral.jl]: This doesn't look right.

File: src/spectral.jl [hunk 2]
```diff
@@ -167,3 +129,134 @@ function incidence_matrix(g::SimpleGraph, T::DataType=Int)
...
+    m::Int
+end
+
+import Base: size, eltype, issym
+size(nbt::Nonbacktracking) = (nbt.m,nbt.m)
+length(nbt::Nonbacktracking) = nbt.m*nbt.m
+eltype(nbt::Nonbacktracking) = Float64
+issym(nbt::Nonbacktracking) = false
+
+function Nonbacktracking(g::SimpleGraph)
```
Comment (sbromberger) [inline src/spectral.jl]: Is there a reason this function is capitalized?

File: src/spectral.jl [hunk 3]
```diff
@@ -167,3 +129,134 @@ function incidence_matrix(g::SimpleGraph, T::DataType=Int)
...
+    if !is_directed(g)
+        for e in edges(g)
+            m += 1
+            edgeidmap[reverse(e)] = m
+        end
+    end
+    return Nonbacktracking(g, edgeidmap,m)
+end
+
+import Base.*
```
Comment (sbromberger) [inline src/spectral.jl]: This import statement doesn't look right.

File: src/spectral.jl [hunk 4]
```diff
@@ -167,3 +129,134 @@ function incidence_matrix(g::SimpleGraph, T::DataType=Int)
...
+Additionally the contract!(vertexspace, nbt, edgespace) method takes vectors represented in
+the domain of B and represents them in the domain of the adjacency matrix of g.
+"""
+type Nonbacktracking{G}
+    g::G
+    edgeidmap::Dict{Edge,Int}
+    m::Int
+end
+
+import Base: size, eltype, issym
```
Comment (sbromberger) [inline src/spectral.jl]: I'm not sure we need this. Check the main imports in LightGraphs.jl?

File: src/spectral.jl (hunk 2 above)
Comment (CarloLucibello) [inline src/spectral.jl]: this is alright, the function is a constructor for the `Nonbacktracking` type

Comment (sbromberger) [inline src/spectral.jl]: Ah, of course. I don't know how I missed the type declaration just above it.

**Comment by CarloLucibello** [comment]
This looks like a welcome performance improvement. Did you tried to compare timing beetween the new and the old version?

**Comment by CarloLucibello** [comment]
Also maybe the Nonbacktracking type and related methods could be moved to GraphMatrices.jl?

Is there a method that converts a GraphMatrices's type to a standard or sparse matrix? If so, should we export it through LightGraphs?

**Comment by jpfairbanks** [comment]
GraphMatrices uses `sparse` to convert to a standard sparse matrix.

**Comment by jpfairbanks** [comment]
I added a benchmark. It started out simple but then I wanted to improve it...

File: src/spectral.jl [hunk 5]
```diff
@@ -167,3 +129,154 @@ function incidence_matrix(g::SimpleGraph, T::DataType=Int)
...
+Additionally the contract!(vertexspace, nbt, edgespace) method takes vectors represented in
+the domain of B and represents them in the domain of the adjacency matrix of g.
+"""
+type Nonbacktracking{G}
+    g::G
+    edgeidmap::Dict{Edge,Int}
+    m::Int
+end
+
+import Base: size, eltype, issym, *, sparse
```
Comment (sbromberger) [inline src/spectral.jl]: Why is this import here?

File: src/spectral.jl [hunk 6]
```diff
@@ -167,3 +129,154 @@ function incidence_matrix(g::SimpleGraph, T::DataType=Int)
...
+            push!(J, u)
+        end
+    end
+    return I,J,1.0
+end
+
+function sparse(nbt::Nonbacktracking)
+    return sparse(coo_sparse(nbt)..., nbt.m,nbt.m)
+end
+
```
Comment (sbromberger) [inline src/spectral.jl]: Better written as `sparse(nbt::Nonbacktracking) = sparse(coo_sparse(nbt)..., nbt.m,nbt.m)` ?

File: src/spectral.jl [hunk 7]
```diff
@@ -167,3 +129,154 @@ function incidence_matrix(g::SimpleGraph, T::DataType=Int)
...
+returns a matrix B, and an edgemap storing the oriented edges' positions in B
+"""
+function non_backtracking_matrix(g::SimpleGraph)
+    # idedgemap = Dict{Int, Edge}()
+    edgeidmap = Dict{Edge, Int}()
+    m = 0
+    for e in edges(g)
+        m += 1
+        edgeidmap[e] = m
+    end
```
Comment (sbromberger) [inline src/spectral.jl]: Why not use

```
for (m,e) in enumerate(edges(g))
...
``` ?

File: src/spectral.jl (hunk 5 above)
Comment (jpfairbanks) [inline src/spectral.jl]: so that it is close to the usage. and importing things twice doesn't hurt.

File: src/spectral.jl (hunk 7 above)
Comment (jpfairbanks) [inline src/spectral.jl]: This is the code that @CarloLucibello wrote.
The next section uses `m` after the loop in order to add in the reverse edges?

Comment (sbromberger) [inline src/spectral.jl]: Ah, ok. I see how that's working.

File: src/spectral.jl (hunk 5 above)
Comment (sbromberger) [inline src/spectral.jl]: The imports should be in LightGraphs.jl so that we know what's going on. The only exception I think we make is in tests.

Comment (jpfairbanks) [inline src/spectral.jl]: I think that in this case (import with the intent to extend) importing near the new definition makes it more clear especially when it is an import from Base.

**Comment by jpfairbanks** [comment]
I squashed commits and I am about to git push -f onto this branch. Then I will merge into master and tag a release.

**Comment by sbromberger** [comment]
Thanks!!!
